pkg/volume/azure_dd/controllerCommon.go

```diff
@@ -0,0 +1,520 @@
+	clientSecret        string
+	managementEndpoint  string
+	tokenEndPoint       string
+	aadResourceEndPoint string
+	aadToken            string
+	expires_on          time.Time
+}
+
+func (c *controllerCommon) isManagedArmVm(storageProfile map[string]interface{}) bool {
+	osDisk := storageProfile["osDisk"].(map[string]interface{})
+
+	_, ok := osDisk["managedDisk"]
+	if ok {
+		return true
+	}
+
+	return false
+
+}
+func (c *controllerCommon) getAttachedDisks(nodeName string) ([]string, error) {
+	var disks []string
+	var vmData interface{}
+	vm, err := c.getArmVm(nodeName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(vm, &vmData); err != nil {
+		return disks, err
+	}
+
+	fragment := vmData.(map[string]interface{})
+
+	props := fragment["properties"].(map[string]interface{})
+	storageProfile := props["storageProfile"].(map[string]interface{})
+	dataDisks, _ := storageProfile["dataDisks"].([]interface{})
+
+	// we silently ignore, if VM does not have the disk attached
+	for _, v := range dataDisks {
+		d := v.(map[string]interface{})
+		if _, ok := d["vhd"]; ok {
+			// this is a blob disk
+			vhdInfo := d["vhd"].(map[string]interface{})
+			vhdUri := vhdInfo["uri"].(string)
+			disks = append(disks, vhdUri)
+		} else {
+			// this is managed disk
+			managedDiskInfo := d["managedDisk"].(map[string]interface{})
+			managedDiskId := managedDiskInfo["id"].(string)
+			disks = append(disks, managedDiskId)
+		}
+	}
+	return disks, nil
+}
+
+// if disk attached returns bool + lun attached to
+func (c *controllerCommon) isDiskAttached(hashedDiskUri, nodeName string, isManaged bool) (bool, int, error) { //attached, lun (string), error
+
+	attached := false
+	lun := -1
+
+	var vmData interface{}
+
+	vm, err := c.getArmVm(nodeName)
+
+	if err != nil {
+		return attached, lun, err
+	}
+
+	if err := json.Unmarshal(vm, &vmData); err != nil {
+		return attached, lun, err
+	}
+
+	fragment := vmData.(map[string]interface{})
+	props := fragment["properties"].(map[string]interface{})
+	storageProfile := props["storageProfile"].(map[string]interface{})
+	dataDisks, _ := storageProfile["dataDisks"].([]interface{})
+
+	for _, v := range dataDisks {
+		d := v.(map[string]interface{})
+		if isManaged {
+			md := d["managedDisk"].(map[string]interface{})
+			currentDiskId := strings.ToLower(md["id"].(string))
+			hashedCurrentDiskId := makeCRC32(currentDiskId)
+			if hashedCurrentDiskId == hashedDiskUri {
+				attached = true
+				lun = int(d["lun"].(float64))
+				break
+			}
+		} else {
+			blobDisk := d["vhd"].(map[string]interface{})
+			blobDiskUri := blobDisk["uri"].(string)
+			hashedBlobDiskUri := makeCRC32(blobDiskUri)
+			if hashedBlobDiskUri == hashedDiskUri {
+				attached = true
+				lun = int(d["lun"].(float64))
+				break
+			}
+		}
+	}
+
+	return attached, lun, nil
+}
+
+// Updates an arm VM based on the payload
+func (c *controllerCommon) updateArmVm(armVmName string, buffer *bytes.Buffer) error {
+
```

Comment:
delete whitespace. (I'm going to stop commenting on this, please do it everywhere)